socranop/installtool.py

```diff
@@ -51,6 +51,97 @@
+          [load] module pydbus: OK
+          [load] module pydbus: ERROR
+        Error: Cannot import module 'pydbus'. Make sure the 'pydbus' package is installed.
+        Traceback (most recent call last):
+          File ...
+        ModuleNotFoundError: No module named 'pydbus'
+
+    by writing the following three liner:
+
+        with Step("load", "module pydbus",
+                  error_msg="Cannot import module 'pydbus'. Make sure the 'pydbus' package is installed."):
+            importlib.import_module("pydbus")  # import must work; discard retval
+    """
+
+    def __init__(self, tag, details, success_word=None, error_msg=None):
+        super(Step, self).__init__()
+        self.tag = tag
+        self.details = details
+        self.error_msg = error_msg
+        self.success_word = success_word
+        self.has_repeated = None
+
+    @staticmethod
+    def __step_start(tag, details):
+        """Standardized output for all 2-part install/uninstall actions (start)"""
+        print_step(tag, details, end=": ")
+
+    @staticmethod
+    def __step_success(success_word=None, postfix=None, **print_kwargs):
+        """Standardized output for all 2-part install/uninstall actions (successful outcome)"""
+        if postfix is None:
+            print(success_word or "OK", **print_kwargs)
+        else:
+            print(success_word or "OK", postfix, **print_kwargs)
+
+    @staticmethod
+    def __step_failure(error_msg, postfix=None):
+        """Standardized output for all 2-part install/uninstall actions (failure case)"""
+        if postfix is None:
+            print("ERROR")
+        else:
+            print("ERROR", postfix)
+        print("Error:", error_msg)
+
+    def set_success_word(self, success_word):
+        self.success_word = success_word
+
+    def try_again(self, attempt, max_attempt):
+        self.has_repeated = (attempt, max_attempt)
+        self.__step_success(
+            "not yet", postfix=f"(attempt {attempt}/{max_attempt})", end="\r"
+        )
+        self.__step_start(self.tag, self.details)
+
+    def __enter__(self):
+        self.__step_start(self.tag, self.details)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if exc_type is None:
+            if self.has_repeated is None:
+                self.__step_success(self.success_word)
+            else:
+                (attempt, max_attempt) = self.has_repeated
+                self.__step_success(
+                    self.success_word,
+                    postfix=f"(successful attempt {attempt}/{max_attempt})",
+                )
+        else:
+            error_msg = self.error_msg or str(exc_value)
+            if self.has_repeated is None:
+                self.__step_failure(error_msg)
+            else:
+                (attempt, max_attempt) = self.has_repeated
+                self.__step_failure(
+                    error_msg, postfix=f"(final attempt {attempt}/{max_attempt})"
+                )
```

Comment:
We should probably always do a `common.debug(exc_value)` just before returning.  Or maybe only debug-print the exception if we didn't use `str(exc_value)` above?

Reply:
Our current `__exit__` method always returns False, so the Python runtime will always re-raise any exception and thus the exception backtrace (which includes str(exc_value) as the last line) will always be available for examination.

Reply:
I think leaving the backtrace available to the user in case it actually happens is an advantage over having to communicate to them to re-run the installtool with --verbose and take a look at that output. It might even give them an idea what the problem is, and how to solve it without needing to interact with the developers.